core/templates/services/user.service.ts

```diff
@@ -76,9 +77,12 @@ angular.module('oppia').factory('UserService', [
           data: newProfileImageDataUrl
         });
       },
+      setReturnUrl: function(newReturnUrl) {
+        returnUrl = newReturnUrl;
+      },
```

Comment:
Why not `getLoginUrlAsync(returnUrl: string | undefined)` then you don't need to have a service variable.

Reply:
When the `newReturnUrl` is allowed as `undefined` it would be fine to call `getLoginUrlAsync()` so no need to change it elsewhere and it makes sense to me to do it this way instead of this.

Reply:
Just to check, do you suggest adding logic in `top-navigation-bar.directive` to detect if the current page is the exploration page and the player is at the last card, then pass in `newReturnUrl`?

I had considered this, and felt that it may be cleaner to have these checks in a player specific directive instead of a general one.

Reply:
Ah I see, we introduce new buttons that are used to sign in but also we want to modify the behaviour of the buttons in the topnavbar so that they preserve also redirect to the correct places after the login, in that case, this approach makes sense.